compiler/damlc/daml-stdlib-src/DA/Internal/Desugar.daml

```diff
@@ -43,10 +43,18 @@ data PostConsuming t = PostConsuming {}
 class HasInterfaceTypeRep i where
   interfaceTypeRep : i -> TypeRep
 
-class HasInterfaceTypeRep i => Implements t i where
+class HasToInterface i t where
   toInterface : t -> i
+
+class HasFromInterface t i where
   fromInterface : i -> Optional t
 
+type Implements t i =
```

Comment:
@sofiafaro-da I took your suggestion to split the class into single-method classes and instead define it as a constraint synonym. 

The only maybe-issue I see is that a constraint `Implements t Interface` (that is, the interface is a concrete type but the template isn't), is just `(HasInterfaceTypeRep Interface, ...)`, which triggers a "simplifiable class constraint" warning. The warning suggests to enable `MonoLocalBinds`, which [happens to be enabled by default in Daml](https://github.com/digital-asset/daml/blob/has-interface-type-rep/compiler/damlc/daml-opts/daml-opts/DA/Daml/Options.hs#L261), so maybe this is a non-issue.

```
Example.hs:1:1: warning: [-Wsimplifiable-class-constraints]
    • The constraint ‘HasInterfaceTypeRep Token’ matches
        instance HasInterfaceTypeRep Token -- Defined in ‘Example’
      This makes type inference for inner bindings fragile;
        either use MonoLocalBinds, or simplify it using the instance
    • In the type signature:
        transferImpl : Implements t Token =>
                       t -> Party -> Update (ContractId Token)
```

Reply:
We don't need to include `HasInterfaceTypeRep i` in the constraint tuple for `Implements t i`. That solves this problem, I think.

Also I think it would be better to move this into a separate PR. Desugaring into multiple classes is a bigger change and it also affects te daml-docs support that @robin-da is working on.

Reply:
^ edited, sorry

Reply:
I'm now trying a slightly different approach, defining `Implements` as a no-method class with all the split classes as superclasses. This way we can deduce `HasInterfaceTypeRep i` from `Implements t i`.

The other reason was that LFConversion uses the `Implements t i` instances to determine which templates implement which interfaces (and also uses the `Implements i i` instances to convert the interfaces). We could rely on one of the split classes instead, but I think this is clearer.

Reply:
I don't think this buys us anything. The reason to have single method typeclasses bundled into a constraint tuple is so we can change them later without breaking data-dependencies. Using a typeclass here instead of a constraint tuple makes it so that ghc won't repackage the constraints automatically if the API changes, so it isn't any better than just having a single `Implements` typeclass.

In our stdlib code, we tend to rely on the methods we specifically need, not on constraint tuples (e.g. look at `exerciseByKey`).